### PR TITLE
Remove Pseudo-Headers From Public API

### PR DESCRIPTION
**Pull request by Daniel-Genkin-MS-2**
Currently we place the pseudo-headers in the `HeaderNames.cs` file which exposes them as public API. This is redundant as the pseudo-headers are all stripped out in internal code prior to passing it up to publicly available layers. Thus, these headers cannot actually be used by end users. This PR marks them as obsolete and adds them to an internal-only file, `PseudoHeaderNames.cs`.

From: https://github.com/dotnet/aspnetcore/issues/42002#issuecomment-1147718108

### CODE REVIEW

File: src/Http/Headers/src/HeaderNames.cs [hunk 1]
```diff
@@ -178,18 +179,21 @@ public static class HeaderNames
...
     /// <summary>Gets the <c>:path</c> HTTP header name.</summary>
+    [Obsolete("For infrastructure use only", false)]
     public static readonly string Path = ":path";
 
     /// <summary>Gets the <c>Pragma</c> HTTP header name.</summary>
     public static readonly string Pragma = "Pragma";
 
     /// <summary>Gets the <c>Protocol</c> HTTP header name.</summary>
+    [Obsolete("For infrastructure use only", false)]
     public static readonly string Protocol = ":protocol";
```
Comment (Tratcher) [inline src/Http/Headers/src/HeaderNames.cs:197]: This one can be removed, it hasn't shipped yet.

File: src/Http/Headers/src/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -4,3 +4,9 @@
 using System.Runtime.CompilerServices;
 
 [assembly: InternalsVisibleTo("Microsoft.Net.Http.Headers.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Server.Kestrel.Core, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (Tratcher) [inline src/Http/Headers/src/Properties/AssemblyInfo.cs:7]: While we do use InternalsVisibleTo for tests, we almost never use it between production assemblies.

It looks like the only component outside of Kestrel.Core that needs these pseudo headers is http2cat? If that's all, http2cat can have its own copy and the main version can move into Kestrel.Core.

**Review by Tratcher — APPROVED** [review]
Let @halter73 review this next week.

File: src/Http/Headers/src/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (Daniel-Genkin-MS-2) [reply src/Http/Headers/src/Properties/AssemblyInfo.cs:7]: Apparently, `Microsoft.Net.Http.Headers.Tests` also needs it, as CI just pointed out. Should I make a copy there as well? I guess we could update the CodeGenerator to make multiple copies and just place each one is the correct directory. Would this be a good solution?

Comment (Tratcher) [reply src/Http/Headers/src/Properties/AssemblyInfo.cs:7]: Where do those tests need them?

Comment (Daniel-Genkin-MS-2) [reply src/Http/Headers/src/Properties/AssemblyInfo.cs:6]: Oh, I misunderstood, `Microsoft.Net.Http.Headers.Tests` doesn't need the pseudo-headers, but the dependency existed prior to this PR.

File: src/Servers/Kestrel/samples/http2cat/PseudoHeaderNames.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Net.Http.Headers;
```
Comment (Tratcher) [inline src/Servers/Kestrel/samples/http2cat/PseudoHeaderNames.cs:4]: Update the namespace to match other h2cat files

File: src/Servers/Kestrel/samples/http2cat/PseudoHeaderNames.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (Tratcher) [inline src/Servers/Kestrel/samples/http2cat/PseudoHeaderNames.cs:1]: Http2cat's code lives over in https://github.com/dotnet/aspnetcore/tree/76ef7a069666e28c8123fa21b70c274191fb23d5/src/Shared/Http2cat

**Review by Tratcher — COMMENTED** [review]
@halter73 will want to review this 😁

File: src/Servers/Kestrel/Core/src/PseudoHeaderNames.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core;
+
+[System.Diagnostics.CodeAnalysis.SuppressMessage("Performance", "CA1802:Use literals where appropriate", Justification = "So referenceEquals works")]
+internal static class PseudoHeaderNames
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/PseudoHeaderNames.cs:7]: Can we please share this with ` src/Shared/Http2cat/PseudoHeaderNames.cs`?

**Review by halter73 — CHANGES_REQUESTED** [review]
Thanks! This looks good.

For APIs that people actually used, we should start using diagnostic IDs for obsoletions. See https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/better-obsoletion.md. I don't think it's necessary for this though because these header names have never been populated in header dictionaries.

File: src/Servers/Kestrel/tools/CodeGenerator/CodeGenerator.csproj [hunk 6]
```diff
@@ -9,6 +9,7 @@
 
   <ItemGroup>
     <Compile Include="$(KestrelSharedSourceRoot)KnownHeaders.cs" />
+    <Compile Include="..\..\Core\src\PseudoHeaderNames.cs" />
```
Comment (Tratcher) [inline src/Servers/Kestrel/tools/CodeGenerator/CodeGenerator.csproj:12]: PseudoHeaderNames is also needed by `src\Servers\HttpSys\test\FunctionalTests\Microsoft.AspNetCore.Server.HttpSys.FunctionalTests.csproj`.

Move it to https://github.com/dotnet/aspnetcore/tree/main/src/Shared

**Review by halter73 — APPROVED** [review]

**Pull request merged**